### PR TITLE
xrootd: init at 5.4.0

### PR DESCRIPTION
**Pull request by ShamrockLee**
###### Motivation for this change
XRootD is a commonly-used data protocol for the high-energy physics, and a missing dependency for ROOT, gfal2, etc..

Cc: @veprbl 
###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by ShamrockLee** [comment]
Trying to make it build on Darwin by replacing libkrb5 with libheimdal.

**Comment by ShamrockLee** [comment]
Which platform does the error occur on? Is it `x86_64-darwin`?

**Comment by veprbl** [comment]
Yes, that's on x86_64-darwin

**Comment by veprbl** [comment]
Why not just use `libkrb5` on darwin?

**Comment by ShamrockLee** [comment]
> Why not just use `libkrb5` on darwin?

There seems to be an evaluation problem shown in the OfBorg aarch64-darwin test caused by 

```Nix
  krb5 = callPackage ../development/libraries/kerberos/krb5.nix {
    inherit (buildPackages.darwin) bootstrap_cmds;
  };
```

in `all-packages.nix`

Not sure if things work on x86_64-darwin.

**Comment by ShamrockLee** [comment]
Updated.

**Comment by veprbl** [comment]
Fails on x86_64-darwin
```
error: builder for '/nix/store/57hjcr4qp26j89ai47wdmd68mxjryrzd-xrootd-5.4.0.drv' failed to produce output path for output 'bin' at '/nix/store/yvhzc29qbrarxxx892fziz5nflnndi8p-xrootd-5.4.0-bin'
```
https://gist.github.com/veprbl/0334679fe712bc57465d8d51d4bd6224

**Comment by ShamrockLee** [comment]
It somehow doesn't respect `CMAKE_BIN_DIR` on Darwin.

**Comment by veprbl** [comment]
Result of `nixpkgs-review pr 160102` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>xrootd</li>
  </ul>
</details>

**Review by veprbl — APPROVED** [review]

**Comment by ShamrockLee** [comment]
Sorry for pushing again. Just found the way to build the testcases.

Would it cause problem to reference the upstream CD's way of calling the tests and "implement" it here?

```Nix
  # These tests are specified in the test procedure of the upstream CD:
  # https://github.com/xrootd/xrootd/blob/master/.github/workflows/build.yml#L90-L98
  installCheckPhase = ''
    runHook preInstallCheck
    TEST_RUNNER="''${bin:-$out}"/bin/test-runner
    XRD_CL_TESTS="$out/lib/libXrdClTests.so"
    "$TEST_RUNNER" "$XRD_CL_TESTS" 'All Tests/UtilsTest/'
    "$TEST_RUNNER" "$XRD_CL_TESTS" 'All Tests/SocketTest/'
    "$TEST_RUNNER" "$XRD_CL_TESTS" 'All Tests/PollerTest/'
    runHook postInstallCheck
  '';
```

Line 90-98 of the [`.github/workflows/build.yml`](https://github.com/xrootd/xrootd/blob/master/.github/workflows/build.yml#L90-L98) upstream:
```yaml
    - name: Run libXrdCl tests
      run: |
        cd ../build/tests
        ./common/test-runner ./XrdClTests/libXrdClTests.so 'All Tests/UtilsTest/'
        ./common/test-runner ./XrdClTests/libXrdClTests.so 'All Tests/SocketTest/'
        ./common/test-runner ./XrdClTests/libXrdClTests.so 'All Tests/PollerTest/'
        # ./common/test-runner ./XrdClTests/libXrdClTests.so 'All Tests/PostMasterTest/'
        # ./common/test-runner ./XrdClTests/libXrdClTests.so 'All Tests/FileSystemTest/'
        # ./common/test-runner ./XrdClTests/libXrdClTests.so 'All Tests/LocalFileHandlerTest/'
```

File: pkgs/tools/networking/xrootd/test-runner.nix [hunk 1]
```diff
@@ -0,0 +1,34 @@
+{ lib
+, cppunit
+, xrootd
+}:
+
+xrootd.overrideAttrs (oldAttrs: {
+  passthru.tests = lib.removeAttrs "test-runner" (oldAttrs.passthru.tests or { });
+
+  cmakeFlags = (oldAttrs.cmakeFlags or [ ]) ++ [
+    "-DENABLE_TESTS=TRUE"
```
Comment (veprbl) [inline pkgs/tools/networking/xrootd/test-runner.nix:10]: Any reason to not always build with tests?

Comment (ShamrockLee) [reply pkgs/tools/networking/xrootd/test-runner.nix:10]: `test-runner.nix` is meant to be imported as `passthru.tests.test-runner` by the main package.

I'm not sure if there would be some infinite recursion when it take itself as a passthru test.

Comment (ShamrockLee) [reply pkgs/tools/networking/xrootd/test-runner.nix:10]: In `xrootd/default.nix`:
```Nix
passthru.tests.test-runner = callPackage ./test-runner.nix { };
```

Comment (veprbl) [reply pkgs/tools/networking/xrootd/test-runner.nix:10]: Let me rephrase. Why not just build the package once with the test enabled?

Comment (ShamrockLee) [reply pkgs/tools/networking/xrootd/test-runner.nix:10]: The menu stated in section [installCheckPhase](https://nixos.org/manual/nixpkgs/stable/#ssec-installCheck-phase)

> It is often better to add tests that are not part of the source distribution to passthru.tests (see [Section 7.1.12, “tests”](https://nixos.org/manual/nixpkgs/stable/#var-meta-tests)). This avoids adding overhead to every build and enables us to run them independently.

Should I enable the test from the cmakeFlags of the original package by default, and just run the test without rebuilding in the testing package?

Comment (veprbl) [reply pkgs/tools/networking/xrootd/test-runner.nix:10]: > Should I enable the test from the cmakeFlags of the original package by default, and just run the test without rebuilding in the testing package?

Yes.

**Comment by ShamrockLee** [comment]
Segfault occurs after reporting that the tests run successfully on `x86_64-darwin` and `aarch64-darwin` Nix-CI.

Not sure if the problem is in the upstream or here.

https://logs.nix.ci/?key=nixos/nixpkgs.160102&attempt_id=10b5c39c-4bef-4edd-ac75-d5a495e214b6

**Comment by veprbl** [comment]
Result of `nixpkgs-review pr 160102` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>xrootd</li>
  </ul>
</details>

**Review by veprbl — APPROVED** [review]
